contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,316 @@
+
+    (uint coverId, uint priceInAsset) = _createCover(owner, productId, payoutAsset, 0, amount, period, stakingPools);
+    require(priceInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(priceInAsset, payoutAsset);
+    return coverId;
+  }
+
+  function createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint8 deniedClaims,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] calldata stakingPools
+  ) external override returns (uint /*coverId*/) {
+    return 0; // TODO: implement
+  }
+
+  function _createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint8 deniedClaims,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint coverId, uint priceInAsset) {
+
+    // convert to NXM amount
+    uint amountToCover = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(payoutAsset));
+    uint totalPrice = 0;
+
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountToCover == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint price) = buyCoverFromPool(stakingPool, productId, amountToCover, period);
+      amountToCover -= coveredAmount;
+      totalPrice += price;
+      stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+    }
+    require(amountToCover == 0, "Not enough available capacity");
+
+    priceInAsset = totalPrice * pool().getTokenPrice(pool().assets(payoutAsset)) / 1e18;
+
+    covers.push(Cover(
+        productId,
+        payoutAsset,
+        deniedClaims,
+        uint96(amount),
+        uint32(block.timestamp + 1),
+        uint32(period),
+        uint96(priceInAsset)
+      ));
+
+    coverId = covers.length - 1;
+    _safeMint(msg.sender, coverId);
+  }
+
+  function buyCoverFromPool(
+    IStakingPool stakingPool,
+    uint24 productId,
+    uint amountToCover,
+    uint32 period
+  ) internal returns (uint, uint) {
+
+    uint availableCapacity = stakingPool.getAvailableCapacity(productId, capacityFactors[productId]);
+
+    uint coveredAmount;
+    if (amountToCover > availableCapacity) {
+      coveredAmount = availableCapacity;
+    } else {
+      coveredAmount = amountToCover;
+    }
+
+    uint capacityFactor = capacityFactors[productId];
+    (uint pricePercentage, uint price) = getPrice(coveredAmount, period, productId, stakingPool);
+    lastPrices[productId][address(stakingPool)] = pricePercentage;
+    lastPriceUpdate[productId][address(stakingPool)] = block.timestamp;
+
+    stakingPool.buyCover(
+      productId,
+      coveredAmount,
+      REWARD_BPS * price / BASIS_PRECISION,
+      period,
+      capacityFactor
+    );
+
+    stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+
+    return (coveredAmount, price);
+  }
+
+  function extendCover(
```

Comment:
we need to walk through all scenarios here and see what happens.

This implementation supports:
- increasing the period
- increasing the amount
- decreasing the amount

It doesn't support:
- decreasing the period